### PR TITLE
k8s: add support for ipFamilies to services

### LINKED ISSUE #13074: Use ipFamilies in k8s 1.20
Related with #12951 

Will be introduced by https://github.com/kubernetes/kubernetes/pull/91824 in K8s 1.20 in December 8th

https://github.com/kubernetes/sig-release/tree/master/releases/release-1.20

### PR DESCRIPTION
**Pull request by fristonio**
Fixes #13074

Service spec:

```
  spec:
    clusterIP: 192.168.128.200
    clusterIPs:
    - 192.168.128.200
    - fe00::22f7
    ipFamilies:
    - IPv4
    - IPv6
    ipFamilyPolicy: RequireDualStack
    ports:
    - port: 80
      protocol: TCP
      targetPort: 80
    selector:
      app: nginx
    sessionAffinity: None
    type: ClusterIP
```

Earlier a service created with dual-stack IP family was not parsed correctly for datapath service in Cilium:

```
cilium service list
ID   Frontend              Service Type   Backend
1    192.168.128.1:443     ClusterIP      1 => 10.128.15.194:6443
2    192.168.128.10:53     ClusterIP      1 => 192.168.64.44:53
                                          2 => 192.168.64.213:53
3    192.168.128.10:9153   ClusterIP      1 => 192.168.64.44:9153
                                          2 => 192.168.64.213:9153
4    192.168.128.200:80    ClusterIP      1 => 192.168.64.205:80
                                          2 => [fc00::b5d]:80
```

After the patch:

```
root@ipfamilies-test-instance:/home/cilium# cilium service list
ID   Frontend              Service Type   Backend
1    192.168.128.1:443     ClusterIP      1 => 10.128.15.194:6443
2    192.168.128.10:53     ClusterIP      1 => 192.168.64.44:53
                                          2 => 192.168.64.213:53
3    192.168.128.10:9153   ClusterIP      1 => 192.168.64.44:9153
                                          2 => 192.168.64.213:9153
4    192.168.128.200:80    ClusterIP      1 => 192.168.64.205:80
5    [fe00::22f7]:80       ClusterIP      1 => [fc00::b5d]:80
```

### CODE REVIEW

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
retest-4.9

File: pkg/k8s/service.go
Comment (aanm) [inline pkg/k8s/service.go:121]: Which guarantees do we have that `svc.Spec.ClusterIPs` is always ordered? If this is set in a different order across updates then it will mean that `func (s *Service) String() string {` will return different values for each k8s update.

File: pkg/k8s/service_cache.go
Comment (aanm) [inline pkg/k8s/service_cache.go]: Why are we picking the first element of `FrontendIPs`?

File: pkg/k8s/service_cache.go
Comment (aanm) [inline pkg/k8s/service_cache.go]: same here: Why are we picking the first element of `FrontendIPs`?

File: pkg/k8s/service_cache.go
Comment (aanm) [inline pkg/k8s/service_cache.go]: same here? ... I'm starting to see a pattern but I still don't understand the why :slightly_smiling_face:

File: pkg/k8s/watchers/watcher.go
Comment (aanm) [inline pkg/k8s/watchers/watcher.go]: some hackity-hack: `isv6 := strings.Contains(ip, ":")` (but leave the `parsedIP := net.ParseIP(ip)` since we will use it :-) )

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/service.go
Comment (fristonio) [reply pkg/k8s/service.go:121]: Is there a reason we want the string output of the service to be consistent? Its just a representation of a service object that we use while logging action related to the service.
For `DeepEquals` we already make sure that the order of the IPs does not matter.

File: pkg/k8s/service_cache.go
Comment (fristonio) [reply pkg/k8s/service_cache.go]: From the comment of the function, this function is used to return a random L3n4Addr that is externally scoped(all frontend IPs are externally scoped). So we pick the first element as it's mostly consistent with the `service.Spec.ClusterIP`.
https://github.com/cilium/cilium/pull/14914/files/35224e9f8741c6d395b7cf754d6454bc6cddff22#diff-054519b4e1e87efe881596a837f8c8d0bf0c52e47bd19270aace9c9f1c86fc18R110-R112

File: pkg/k8s/service_cache.go
Comment (fristonio) [reply pkg/k8s/service_cache.go]: Here we should not pick a single IP address, rather we should return all the valid frontend IPs associated with the service. I will update the PR.

File: pkg/k8s/watchers/watcher.go
Comment (fristonio) [reply pkg/k8s/watchers/watcher.go]: I was thinking of making the use of `IsIPv6` consistent, so its easy to understand simply looking at the function what it informs.

File: pkg/k8s/service_cache.go
Comment (fristonio) [reply pkg/k8s/service_cache.go]: Same as above.

File: pkg/k8s/service.go
Comment (aanm) [reply pkg/k8s/service.go:121]: Yes, the `String()` of `pkg/k8s/service.go:Service` is used has map keys so it needs to be consistent.

File: pkg/k8s/service_cache.go
Comment (aanm) [reply pkg/k8s/service_cache.go]: Oh... interesting... In that case perhaps it would be better to only return a for which the IP family is supported by Cilium. For example, if Cilium is running with `ipv4-enable: false` then we need to make sure `svc.FrontendIPs[0]` is *not* a v4 address.

File: pkg/k8s/service_cache.go
Comment (aanm) [reply pkg/k8s/service_cache.go]: Same answer as before

File: pkg/k8s/service_cache.go
Comment (fristonio) [reply pkg/k8s/service_cache.go]: Yeah, that totally makes sense. Will update the PR.

File: pkg/ip/ip.go
Comment (aanm) [inline pkg/ip/ip.go]: small nit `ip != nil` is not necessary since `ip.To4()` is checking for that as well

File: pkg/ip/ip.go
Comment (aanm) [inline pkg/ip/ip.go:812]: small nit `ip != nil` is not necessary since `ip.To4()` is checking for that as well

File: pkg/ip/ip.go
Comment (aanm) [inline pkg/ip/ip.go]: Usually we should return an object like it is being done in here but golang has `sort.Strings()` where it doesn't return an ordered slice so maybe we can do the same to avoid memory allocations for the returned slice.

File: pkg/k8s/service.go
Comment (aanm) [inline pkg/k8s/service.go:152]: My suggestion would be to order the `clusterIPs` here to avoid ordering them every time `String()` is executed.

File: pkg/k8s/service.go
Comment (aanm) [inline pkg/k8s/service.go:449]: Or better yet, ordering them here.

File: pkg/k8s/service.go
Comment (aanm) [inline pkg/k8s/service.go:537]: and here

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/service.go
Comment (fristonio) [reply pkg/k8s/service.go:449]: Done

File: pkg/k8s/service.go
Comment (fristonio) [reply pkg/k8s/service.go:537]: Done

**Review by aanm — APPROVED** [review]

File: pkg/ip/ip.go
Comment (fristonio) [reply pkg/ip/ip.go:812]: For `IPv6` we need the `ip != nil` because if the `net.IP` passed as the argument is nil the function will return true instead of false.

**Comment by fristonio** [comment]
test-me-please

File: pkg/ip/ip.go
Comment (christarazi) [inline pkg/ip/ip.go:854]: I see you've implemented both variants for checking IP list equality, sorted and unsorted. The latter will simply sort the lists anyway. I guess the assumption being made here is that the caller should know if the IP lists they're passing into these functions are either sorted and unsorted. I'm curious as to why you chose to leave it up to the caller? Should they know if their IP lists are (un)sorted?

Seems to me that if we already sort the lists, then we might as well just have `AreIPListsEqual()` where we sort them under-the-hood, without requiring the caller to know if they are already sorted or not. WDYT?

File: pkg/k8s/service.go
Comment (christarazi) [inline pkg/k8s/service.go:121]: I think we can simplify this block:

```suggestion
	clusterIPs := make([]net.IP, 0, len(svc.Spec.ClusterIPs))
	if clsIP := net.ParseIP(svc.Spec.ClusterIP); clsIP != nil {
		clusterIPs = append(clusterIPs, clsIP)
	}
	// Here we assume that the value of .spec.ClusterIPs[0] is same as that of the .spec.clusterIP
	// or else Kubernetes will reject the service with validation error.
	for _, ip := range svc.Spec.ClusterIPs {
```

File: pkg/k8s/service.go
Comment (christarazi) [inline pkg/k8s/service.go:532]:
```
for i, ip := range svc.Frontends {
 	feIPs[i] = net.ParseIP(ipStr) 
}
```

File: pkg/k8s/service.go
Comment (christarazi) [inline pkg/k8s/service.go:577]: Same as above

File: pkg/k8s/service_test.go
Comment (christarazi) [inline pkg/k8s/service_test.go]: Usually we specify the name of the field when making table-driven tests just for the sake of readability

File: test/helpers/kubectl.go
Comment (christarazi) [inline test/helpers/kubectl.go]: We can replace the `Sprintf` with a pretty simple string concatenation:

```
"service " + service
```

(inspired by https://github.com/cilium/cilium/pull/15089)

File: test/helpers/utils.go
Comment (christarazi) [inline test/helpers/utils.go]: Nit:

```
switch GetCurrentK8SEnv() {
```

File: test/helpers/utils.go
Comment (christarazi) [inline test/helpers/utils.go]: I thought I recall seeing some existing code where we can do version "arithmetic" like `<1.20`? Anyway this is pretty minor, up to you

**Review by christarazi — CHANGES_REQUESTED** [review]
Nice! Some comments below.

File: pkg/ip/ip.go
Comment (kkourt) [reply pkg/ip/ip.go:854]: [General comments, no idea how callers of this work]

The issue with sorting the lists under the hood is that previously kept indices will become invalid. I would prefer keeping the code as is. If it is decided to sort things under the hood, please at least rename  `AreIPListsEqual()` to something that indicates that the order might change.

File: pkg/k8s/service_cache.go
Comment (kkourt) [inline pkg/k8s/service_cache.go:122]: As it is nicely documented elsewhere:
> `GetIPFromListByFamily` returns a single IP address of the provided family from a list

Is it possible to set multiple IPs of the same family in the `.FrontendIPs` field?
If yes, should we have a selection here that will not always select the same?

**Review by kkourt — COMMENTED** [review]
Some first comments.

File: pkg/ip/ip.go
Comment (christarazi) [reply pkg/ip/ip.go:854]: That's a good point ^

File: pkg/k8s/service.go
Comment (aanm) [reply pkg/k8s/service.go:532]: @christarazi `svc.Frontends` is a map, so `i, ip` will return the "key" and the "value".

File: pkg/k8s/service.go
Comment (aanm) [reply pkg/k8s/service.go:577]: same as above

File: pkg/k8s/service.go
Comment (aanm) [reply pkg/k8s/service.go:121]: That will cause `clusterIPs` to have duplicated IP addresses. (`svc.Spec.ClusterIP` == `svc.Spec.ClusterIPs[0]`)

File: pkg/ip/ip.go
Comment (fristonio) [reply pkg/ip/ip.go:854]: I agree with @kkourt 
The idea was that it is the responsibility of the caller to know what kind of equality should be checked for.

File: test/helpers/kubectl.go
Comment (fristonio) [reply test/helpers/kubectl.go]: Good point, will do this. :+1:

File: pkg/k8s/service.go
Comment (christarazi) [reply pkg/k8s/service.go:532]: Thanks, clearly I wasn't paying too close of attention :facepalm:

File: pkg/k8s/service.go
Comment (christarazi) [reply pkg/k8s/service.go:121]: Ah good catch

File: pkg/ip/ip.go
Comment (brb) [inline pkg/ip/ip.go]: Technically it's possible to return a nil address. `IsIPv4()` might return `false` for a nil `net.IP`. I'd extend the check by appending `|| !v4Family && IsIPv6(..)`.

File: pkg/k8s/service_cache.go
Comment (brb) [reply pkg/k8s/service_cache.go:122]: > Is it possible to set multiple IPs of the same family in the .FrontendIPs field?
I don't think so. Only a single ClusterIP (VIP) per family should be allowed.

**Review by brb — CHANGES_REQUESTED** [review]
Thanks, just one nit which needs to be addressed.

Also, I assume that it is OK to keep the same logic for deciding whether to create IPv4 or IPv6 NodePort service in `ParseService()` function?

File: pkg/ip/ip.go
Comment (fristonio) [reply pkg/ip/ip.go]: Makes sense to me. Will update the PR.

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
test-gke

**Review by christarazi — APPROVED** [review]

**Review by brb — APPROVED** [review]
:rocket:

**Review by kkourt — APPROVED** [review]

**Review by joestringer — APPROVED** [review]
I'm not quite sure why GitHub pulled in janitor codeowners for review, all of the files seem to have other codeowners already. Reviewers seem otherwise happy so I see no need to interdict.

**Pull request merged**
